### PR TITLE
Implement Test Operator Service

### PR DESCRIPTION
**Pull request by Spikhalskiy**
## What was changed

Implemented Test Operator Service
Refactored Test Server to support pluggable services
Created new public Test Server API

Issue #617

### CODE REVIEW

File: temporal-test-server/src/main/java/io/temporal/internal/testservice/TestVisibilityStoreImpl.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+public class TestVisibilityStoreImpl implements TestVisibilityStore {
+
+  private static final String TEST_KEY_STRING = "CustomStringField";
+  private static final String TEST_KEY_INTEGER = "CustomIntField";
+  private static final String TEST_KEY_DATE_TIME = "CustomDatetimeField";
+  private static final String TEST_KEY_DOUBLE = "CustomDoubleField";
+  private static final String TEST_KEY_BOOL = "CustomBoolField";
+
+  private final Map<String, IndexedValueType> searchAttributes =
+      new HashMap<String, IndexedValueType>() {
```
Comment (mmcshane) [inline temporal-test-server/src/main/java/io/temporal/internal/testservice/TestVisibilityStoreImpl.java:36]: Worthwhile to throw a lock around access to this? Or make it a synchronized map?

File: temporal-test-server/src/main/java/io/temporal/internal/testservice/TestVisibilityStoreImpl.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      throw Status.INVALID_ARGUMENT
+          .withDescription("Unable to read search attribute type: " + type)
+          .asRuntimeException();
+    }
+    searchAttributes.put(name, type);
+  }
+
+  @Override
+  public Map<String, IndexedValueType> getRegisteredSearchAttributes() {
+    return searchAttributes;
```
Comment (mmcshane) [inline temporal-test-server/src/main/java/io/temporal/internal/testservice/TestVisibilityStoreImpl.java:58]: Maybe return a copy here? Or is it intentional that the registry can be mutated through this accessor?

**Review by mmcshane — COMMENTED** [review]
Comment on locking may or may not be needed - I'm 100% on the expected usage

Confused though as to why we're not sending `sleep` to the out-of-proc server?

File: temporal-testing/src/main/java/io/temporal/testing/TestWorkflowEnvironmentInternal.java [hunk 3]
```diff
@@ -112,32 +114,58 @@ public Worker newWorker(String taskQueue, WorkerOptions options) {
...
   }
 
   @Override
   public void sleep(Duration duration) {
-    service.sleep(duration);
+    if (service != null) {
+      service.sleep(duration);
+    } else {
+      try {
+        Thread.sleep(duration.toMillis());
```
[the remark this answers is not in the document]
Comment (Spikhalskiy) [reply temporal-testing/src/main/java/io/temporal/testing/TestWorkflowEnvironmentInternal.java:146]: We don't support time skipping for external not in-process server. So, the only time skipping for them right now is sleeping. I will change it when we expose time skipping as gRPC service.

[the remark this answers is not in the document]
Comment (mmcshane) [reply temporal-testing/src/main/java/io/temporal/testing/TestWorkflowEnvironmentInternal.java:146]: Got it. Thanks for explaining.

**Comment by Spikhalskiy** [comment]
@mmcshane
What can be an out-of-process test server?
Dockerized Service. Doesn't support time skipping.
Test Service. Doesn't expose time skipping API right now. And who will be using the out of process test server with java if there is a java version? :)

**Pull request merged**